examples/02_intermediate/wikimedia.ipynb

```diff
@@ -0,0 +1,513 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "from datetime import datetime\n",
+    "\n",
+    "URL = 'https://stream.wikimedia.org/v2/stream/recentchange'\n",
+    "\n",
+    "# Initial data\n",
+    "data = pd.DataFrame({\n",
+    "    \"servername\": [\"\"],\n",
+    "    \"time\": [datetime.utcnow()],\n",
+    "    \"event\": [\"\"],\n",
+    "    \"servername_count\": [0],\n",
+    "})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fcdde9db-2273-4ac9-998c-d4f0d0b8c5a4",
+   "metadata": {},
+   "source": [
+    "Next, we'll set up an initial (empty) visualization widget from the Perspective library so we can visualize the streaming data later. (Note that if you are running this example locally on JupyterLab, you may need to restart the Jupyter server after installing the Perspective library in order to visualize the widget.)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ad33d769-c1a2-439b-be02-7abc7faccfdd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from perspective import PerspectiveWidget, Table, Plugin\n",
+    "\n",
+    "table = Table(data, index=\"servername\")\n",
+    "view = table.view()\n",
+    "df = view.to_df()\n",
+    "widget = PerspectiveWidget(table, plugin=Plugin.XBAR, group_by=[\"servername\"], columns=[\"servername_count\"])\n",
+    "widget"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "57481e87-c37b-4c89-a05e-7fc93abf8a56",
+   "metadata": {},
+   "source": [
+    "Now, we will use CSP to build our EventSource adapter and update the Perspective widget we defined above with real-time events from the stream.\n",
+    "\n",
+    "For this, we will use a [PushInputAdapter](https://github.com/Point72/csp/wiki/Write-Realtime-Input-Adapters#pushinputadapter---python). This `PushInputAdapter` that you define will be used as the _runtime implementation_, and you also need to define a _graph- time_ representation of the time series edge by using `py_push_adapter_def`. \n",
+    "\n",
+    "When running the cell below, you should see the widget updated in real time with each new event, indexed by the server where the update happened."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5667cea7-43c5-471c-9859-757c6ac18be2",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import threading\n",
+    "\n",
+    "from sseclient import SSEClient as EventSource, Event\n",
+    "\n",
+    "import csp\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "# Define the runtime implementation of our adapter\n",
+    "class FetchWikiDataAdapter(PushInputAdapter):\n",
+    "    def __init__(self, widget):\n",
+    "        self._thread = None\n",
+    "        self._running = False\n",
+    "        self.widget = widget\n",
```

Comment:
This isnt a great demonstration of csp.  The adapter shouldnt be updating the widget directly at all.  It should be responsible for one thing, which is streaming data from wikimedia.
So I think the way this should be structured is that you define a WikiData struct that has the fields that you need ( servername, edit_time, event string )

The adapter will create an instance of WikiData and push that out.
A separate node can be define that consumes ts[WikiData] and takes a widget to update.  When WikiData ticks, it applies it to the widget. You cna also csp.print() the ts[WikiData] and remove the direct print in the adapter code

Reply:
Yeah, that makes sense to me. Since I was new to Perspective as well, I wasn't sure what the best option was. Thanks for the suggestions, I'll fix it 👍🏻

Reply:
Refactored; I think this is addressing your comment, let me know otherwise.